src/whylogs/core/metrics/confusion_matrix.py

```diff
@@ -0,0 +1,148 @@
+from typing import List, Union
+
+import numpy as np
+from sklearn.utils.multiclass import type_of_target
+
+from whylogs.proto import ScoreMatrixMessage
+from whylogs.core.statistics import NumberTracker
+
+
+SUPPORTED_TYPES = ("binary", "multiclass")
+
+
+class ConfusionMatrix:
+
+    """
+    Confusion Matrix Class to hold labels and matrix data
+
+    Attributes:
+        confusion_matrix (nd.array): Description
+        labels (List[str]): Description
+    """
+
+    def __init__(self, labels: List[str] = None,
+                 prediction_field=None,
+                 target_field=None,
+                 score_field=None):
+        self.prediction_field = prediction_field
+        self.target_field = target_field
+        self.score_field = score_field
+        if labels:
+            self.labels = sorted(labels)
+            num_labels = len(self.labels)
+            self.confusion_matrix = np.empty(
+                [num_labels, num_labels], dtype=object)
+            for each_ind_i in range(num_labels):
+                for each_ind_j in range(num_labels):
+                    self.confusion_matrix[each_ind_i,
+                                          each_ind_j] = NumberTracker()
+        else:
+            self.labels = None
+            self.confusion_matrix = None
+
+    def add(self, predictions: List[Union[str, int, bool]],
+            targets: List[Union[str, int, bool]],
+            scores: List[float]):
+        """
+        add. predictions and targets to confusion matrix with scores
+
+        Args:
+            predictions (TYPE): Description
+            targets (TYPE): Description
+            scores (TYPE): Description
+
+        Raises:
+            ValueError: Description
+        """
+
+        tgt_type = type_of_target(targets)
+        if tgt_type not in ("binary", "multiclass"):
+            raise NotImplementedError("target type not supported yet")
+
+        if not isinstance(targets, list):
+            targets = [targets]
+        if not isinstance(predictions, list):
+            predictions = [predictions]
+
+        if scores is None:
+            scores = [1.0 for _ in range(len(targets))]
+
+        if len(targets) != len(predictions):
+            raise ValueError(
+                "both targets and predictions need to have the same length")
+
+        targets_indx = enconde_to_integers(targets, self.labels)
+        prediction_indx = enconde_to_integers(predictions, self.labels)
+
+        for ind in range(len(predictions)):
+            self.confusion_matrix[prediction_indx[ind],
+                                  targets_indx[ind]].track(scores[ind])
+
+    def merge(self, other_CM):
+        """
+        merge two seperate confusion matrix which may or may not overlap in labels
+
+        Args:
+            other_CM (TYPE): Description
+
+        Returns:
+            TYPE: Description
+        """
+
+        if self.labels is None or self.labels == []:
+            return other_CM
```

Comment:
The `merge` method in other classes ALWAYS return a new object:

example: https://github.com/whylabs/whylogs-python/blob/mainline/src/whylogs/core/columnprofile.py#L135